src/test/controllerscriptenginelegacy_test.cpp

```diff
@@ -57,13 +79,104 @@ class ControllerScriptEngineLegacyTest : public ControllerScriptEngineLegacy, pu
+        // but it does a lot of local disk / settings setup.
+        auto pChannelHandleFactory = std::make_shared<ChannelHandleFactory>();
+        m_pEffectsManager = std::make_shared<EffectsManager>(config(), pChannelHandleFactory);
+        m_pEngine = std::make_shared<EngineMixer>(
+                config(),
+                "[Master]",
+                m_pEffectsManager.get(),
+                pChannelHandleFactory,
+                true);
+        m_pSoundManager = std::make_shared<SoundManager>(config(), m_pEngine.get());
+        m_pControlIndicatorTimer = std::make_shared<mixxx::ControlIndicatorTimer>(nullptr);
+        m_pEngine->registerNonEngineChannelSoundIO(gsl::make_not_null(m_pSoundManager.get()));
+
+        CoverArtCache::createInstance();
+
+        m_pPlayerManager = std::make_shared<PlayerManager>(config(),
+                m_pSoundManager.get(),
+                m_pEffectsManager.get(),
+                m_pEngine.get());
+
+        m_pPlayerManager->addConfiguredDecks();
+        m_pPlayerManager->addSampler();
+        PlayerInfo::create();
+        m_pEffectsManager->setup();
+
+        const auto dbConnection = mixxx::DbConnectionPooled(dbConnectionPooler());
+        if (!MixxxDb::initDatabaseSchema(dbConnection)) {
+            exit(1);
+        }
+
+        m_pTrackCollectionManager = std::make_shared<TrackCollectionManager>(
+                nullptr,
+                config(),
+                dbConnectionPooler(),
+                [](Track* pTrack) { delete pTrack; });
+
+        m_pRecordingManager = std::make_shared<RecordingManager>(config(), m_pEngine.get());
+        m_pLibrary = std::make_shared<Library>(
+                nullptr,
+                config(),
+                dbConnectionPooler(),
+                m_pTrackCollectionManager.get(),
+                m_pPlayerManager.get(),
+                m_pRecordingManager.get());
+
+        m_pPlayerManager->bindToLibrary(m_pLibrary.get());
+        ControllerScriptEngineBase::registerPlayerManager(m_pPlayerManager);
+        ControllerScriptEngineBase::registerTrackCollectionManager(m_pTrackCollectionManager);
+    }
+
+    void loadTrackSync(const QString& trackLocation) {
+        TrackPointer pTrack1 = m_pTrackCollectionManager->getOrAddTrack(
+                TrackRef::fromFilePath(getTestDir().filePath(trackLocation)));
+        auto* deck = m_pPlayerManager->getDeck(1);
+        deck->slotLoadTrack(pTrack1,
+#ifdef __STEM__
+                mixxx::StemChannelSelection(),
+#endif
+                false);
+        m_pEngine->process(1024);
+        while (!deck->getEngineDeck()->getEngineBuffer()->isTrackLoaded()) {
+            QTest::qSleep(100);
+        }
+        processEvents();
+    }
+
+    void unloadTrackSync() {
+        auto* deck = m_pPlayerManager->getDeck(1);
+        deck->slotEjectTrack(1.0);
+        while (deck->getEngineDeck()->getEngineBuffer()->isTrackLoaded()) {
+            QTest::qSleep(100);
+        }
```

Comment:
Sadly, this doesn't work. The test just hangs. If I execute this in the debugger with breakpoints, `deck->slotEjectTrack(1.0)` correctly ejects the track and the test passes. So I'm not correctly waiting here. There's a concurrency problem. I'd accept any help on how to solve this.

Reply:
Did you try to eject from JavaScript using `engine.setValue('[Channel1]', 'eject', 1);` ?

Reply:
Alright, deal.

Reply:
Sadly no, that still hangs or not even eject, depending on the solution. I tried a number of combinations around this.

Reply:
You probably need to call it twice, as in other tests: https://github.com/mixxxdj/mixxx/blob/14c19a55b9ab0c4bde94d7f61033206081dd698d/src/test/controlobjectscripttest.cpp#L85-L91